File: src/Scripting/Lua/patterns/mediator.lua
local events = {}

local Mediator = {}
Mediator.__index = Mediator

function Mediator.new()
  return setmetatable({ colleagues = {} }, Mediator)
end

function Mediator:register(name, receive)
  self.colleagues[name] = receive
end

function Mediator:send(sender, recipient, message)
  local receive = self.colleagues[recipient]
  assert(receive, "unknown colleague: " .. recipient)
  receive(sender, message)
end

local mediator = Mediator.new()
mediator:register("inventory", function(sender, message)
  table.insert(events, "inventory<-" .. sender .. ":" .. message)
end)
mediator:register("payment", function(sender, message)
  table.insert(events, "payment<-" .. sender .. ":" .. message)
end)

local payment = function(message)
  mediator:send("payment", "inventory", message)
end

local inventory = function(message)
  mediator:send("inventory", "payment", message)
end

payment("paid")
inventory("reserved")

assert(table.concat(events, ",") == "inventory<-payment:paid,payment<-inventory:reserved")
local ok = pcall(function()
  mediator:send("payment", "unknown", "ignored")
end)
assert(not ok)

return true

File: src/Scripting/PHP/patterns/mediator.php
<?php declare(strict_types=1);

final class CheckoutMediator
{
    /** @var array<string, Closure(string, string): void> */
    private array $colleagues = [];

    public function register(string $name, Closure $receive): void
    {
        $this->colleagues[$name] = $receive;
    }

    public function send(string $sender, string $recipient, string $message): void
    {
        if (!isset($this->colleagues[$recipient])) {
            throw new RuntimeException("unknown colleague: $recipient");
        }

        ($this->colleagues[$recipient])($sender, $message);
    }
}

$events = [];
$mediator = new CheckoutMediator();
$mediator->register('inventory', static function (string $sender, string $message) use (&$events): void {
    $events[] = "inventory<-$sender:$message";
});
$mediator->register('payment', static function (string $sender, string $message) use (&$events): void {
    $events[] = "payment<-$sender:$message";
});

$payment = static fn(string $message) => $mediator->send('payment', 'inventory', $message);
$inventory = static fn(string $message) => $mediator->send('inventory', 'payment', $message);

$payment('paid');
$inventory('reserved');

if ($events !== ['inventory<-payment:paid', 'payment<-inventory:reserved']) {
    throw new RuntimeException('Mediator did not coordinate colleagues.');
}

try {
    $mediator->send('payment', 'unknown', 'ignored');
    throw new RuntimeException('Unknown colleague unexpectedly accepted.');
} catch (RuntimeException $error) {
    if (!str_starts_with($error->getMessage(), 'unknown colleague:')) {
        throw $error;
    }
}
